File: src/lsp/actions/test_server.py
"""Framed LSP peer for code-action integration tests; not a language server."""
import json
from pathlib import Path
import sys
from urllib.parse import unquote, urlparse

MODE = sys.argv[1]
SOURCE = None
VERSIONS = []
SEQUENCE = 0


def read():
    length = None
    total = 0
    while True:
        line = sys.stdin.buffer.readline(4097)
        if not line:
            return None
        total += len(line)
        if total > 16384 or len(line) > 4096:
            raise ValueError("oversized test frame headers")
        if line == b"\r\n":
            break
        key, value = line.split(b":", 1)
        if key.lower() == b"content-length":
            length = int(value)
    if length is None or not 0 < length <= 4 * 1024 * 1024:
        raise ValueError("invalid test frame size")
    body = sys.stdin.buffer.read(length)
    if len(body) != length:
        raise ValueError("truncated test frame")
    return json.loads(body)


def send(message):
    payload = json.dumps(dict(jsonrpc="2.0", **message)).encode()
    sys.stdout.buffer.write(b"Content-Length: %d\r\n\r\n" % len(payload) + payload)
    sys.stdout.buffer.flush()


def reply(request, result):
    send({"id": request["id"], "result": result})


def edit(uri, text="fixed", end=3):
    return {"changes": {uri: [{"range": {"start": {"line": 0, "character": 0},
                                         "end": {"line": 0, "character": end}},
                              "newText": text}]}}


def server_edit(uri):
    global SEQUENCE
    SEQUENCE += 1
    request_id = "apply-%d" % SEQUENCE
    send({"id": request_id, "method": "workspace/applyEdit", "params": {"edit": edit(uri)}})
    while True:
        response = read()
        if response is None:
            raise EOFError("client left during server edit")
        if response.get("id") == request_id:
            return response["result"]
        if "method" in response and "id" not in response:
            continue
        raise ValueError("unexpected request while awaiting server edit acknowledgement")


def main():
    global SOURCE
    while True:
        request = read()
        if request is None:
            return
        method = request.get("method")
        params = request.get("params", {})
        if method:
            with Path("requests.jsonl").open("a", encoding="utf-8") as log:
                log.write(json.dumps({"method": method, "params": params}) + "\n")
        if method == "initialize":
            assert params["capabilities"]["textDocument"]["codeAction"]["dataSupport"]
            reply(request, {"capabilities": {"textDocumentSync": 1,
                  "codeActionProvider": {"resolveProvider": True},
                  "executeCommandProvider": {"commands": ["test.finish"]}}})
        elif method in ("textDocument/didOpen", "textDocument/didChange"):
            VERSIONS.append(params["textDocument"]["version"])
        elif method == "textDocument/codeAction":
            SOURCE = params["textDocument"]["uri"]
            action = {"title": "Finish refactoring", "kind": "refactor", "data": {"uri": SOURCE}}
            if MODE == "disabled":
                action["disabled"] = {"reason": "not applicable"}
            reply(request, [action])
        elif method == "codeAction/resolve":
            resolved = dict(params, edit=edit(params["data"]["uri"]),
                            command={"title": "Finish", "command": "test.finish", "arguments": ["literal"]})
            if MODE == "changed":
                resolved["title"] = "Another action"
            if MODE == "malformed":
                resolved["command"]["arguments"] = {"not": "an array"}
            reply(request, resolved)
        elif method == "workspace/executeCommand":
            assert Path(unquote(urlparse(SOURCE).path)).read_text() == "fixed\n", "command preceded edit"
            Path("command-started").write_text("started", encoding="ascii")
            if MODE == "stall":
                continue
            if MODE == "error":
                send({"id": request["id"], "error": {"code": -32801, "message": "command failed after the edit"}})
                continue
            result = None
            if MODE == "callback":
                result = server_edit(Path("sibling.lspfixture").resolve().as_uri())
            elif MODE == "outside":
                result = server_edit((Path.cwd().parent / "outside.lspfixture").as_uri())
            reply(request, result)
        elif method in ("test/unsolicited", "test/lateEdit"):
            result = server_edit(Path("sibling.lspfixture").resolve().as_uri())
            reply(request, result)
        elif method == "test/versions":
            reply(request, VERSIONS)
        elif method == "shutdown":
            reply(request, None)
        elif method == "exit":
            return
        elif "id" in request:
            send({"id": request["id"], "error": {"code": -32601, "message": "unknown fixture method"}})


if __name__ == "__main__":
    main()

File: src/lsp/client/test_server.rs
//! Real framed-stdio peer for native client protocol tests. No provider keys,
//! ports or external language-server installations are required.

use std::path::Path;
use std::process::{Command, Stdio};
use std::time::Duration;

use asupersync::runtime::{Runtime, RuntimeBuilder};
use serde_json::{Value, json};

use super::LspClient;

const SERVER: &str = r#"
import json, sys
capabilities = json.loads(sys.argv[1])
frames, replies, held = [], {}, []
def send(value):
    body = json.dumps(value, separators=(',', ':')).encode('utf-8')
    sys.stdout.buffer.write(('Content-Length: %d\r\n\r\n' % len(body)).encode('ascii') + body)
    sys.stdout.buffer.flush()
while True:
    headers = {}
    while True:
        line = sys.stdin.buffer.readline()
        if not line:
            sys.exit(0)
        if line in (b'\r\n', b'\n'):
            break
        key, value = line.decode('ascii').split(':', 1)
        headers[key.lower()] = value.strip()
    body = sys.stdin.buffer.read(int(headers['content-length']))
    message = json.loads(body)
    frames.append(message)
    method, params = message.get('method'), message.get('params') or {}
    if method == 'test/release':
        request_id = held.pop(0)
        send({'jsonrpc':'2.0', 'id':request_id, 'result':params['result']})
        continue
    if 'id' not in message:
        continue
    response = {'jsonrpc':'2.0', 'id':message['id']}
    if method == 'initialize':
        response['result'] = {'capabilities':capabilities}
    elif method == 'test/frames':
        response['result'] = frames
    elif method == 'test/configure':
        replies.update(params['replies'])
        response['result'] = None
    elif method == 'test/hang':
        held.append(message['id'])
        continue
    elif method == 'test/success':
        response['result'] = {'ok':True}
    elif replies.get(method):
        reply = replies[method].pop(0)
        if reply.get('hold'):
            held.append(message['id'])
            continue
        response.update(reply)
    else:
        response['error'] = {'code':-32601, 'message':'test peer has no response configured'}
    send(response)
"#;

pub(super) struct Fixture {
    pub(super) client: LspClient,
    pub(super) runtime: Runtime,
}

impl Fixture {
    pub(super) fn connect(root: &Path, capabilities: Value) -> Option<Self> {
        let python = ["python3", "python"].into_iter().find(|program| {
            Command::new(program).arg("--version")
                .stdout(Stdio::null()).stderr(Stdio::null())
                .status().is_ok_and(|status| status.success())
        });
        let Some(python) = python else {
            eprintln!("SKIP LSP stdio protocol fixture: Python is not installed");
            return None;
        };
        let runtime = RuntimeBuilder::current_thread().build().expect("runtime");
        let args = vec!["-u".to_string(), "-c".to_string(), SERVER.to_string(), capabilities.to_string()];
        let client = runtime.block_on(LspClient::connect(
            python, &args, &[], root, None, Duration::from_secs(5),
        )).expect("initialize real stdio peer");
        Some(Self { client, runtime })
    }

    pub(super) fn frames(&self) -> Vec<Value> {
        self.runtime.block_on(self.client.call(
            "test/frames", json!({}), Duration::from_secs(5),
        )).expect("protocol frames").as_array().expect("frame array").clone()
    }

    pub(super) fn configure(&self, replies: Value) {
        self.runtime.block_on(self.client.call(
            "test/configure", json!({"replies":replies}), Duration::from_secs(5),
        )).expect("configure protocol replies");
    }
}
